### PR TITLE
Correctly handle typing stream id resetting

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/replication/resource.py [hunk 1]
```diff
@@ -274,11 +274,18 @@ def presence(self, writer, current_token, request_streams):
...
 
         request_typing = request_streams.get("typing")
 
         if request_typing is not None:
+            # If they have a higher token than current max, we can assume that
+            # they had been talking to a previous instance of the master. Since
+            # we reset the token on restart, the best (but hacky) thing we can
+            # do is to simply resend down all the typing notifications.
+            if request_typing > current_position:
+                request_typing = 0
```
Comment (NegativeMjark) [inline synapse/replication/resource.py]: Does this cause the readers to reset their typing_stream token? If not then they will *keep* requesting from replication with the large token, meaning that every request will return all typing updates?

Comment (erikjohnston) [inline synapse/replication/resource.py]: AFAICT slaves always update based on the `position` key

Comment (erikjohnston) [inline synapse/replication/resource.py]: In particular, there is nowhere in the slaves where do a `stream_token = max(last_token, new_token)`, we always just use what was given to us in the response.

Comment (NegativeMjark) [inline synapse/replication/resource.py]: https://github.com/matrix-org/synapse/blob/master/synapse/replication/slave/storage/_slaved_id_tracker.py

Comment (erikjohnston) [inline synapse/replication/resource.py]: Ugh. Also, doesn't apply to typing?

Comment (NegativeMjark) [inline synapse/replication/resource.py]: Oh quite possibly. I'm possibly misremembering how the slave works.

Comment (NegativeMjark) [inline synapse/replication/resource.py]: Maybe add a comment to Typing thing in synapse/app/synchrotron.py to say that the current behaviour is important.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**